File: app/src/main/java/org/wikipedia/language/AcceptLanguageUtil.java
package org.wikipedia.language;

import androidx.annotation.NonNull;

import java.util.Locale;

public final class AcceptLanguageUtil {
    private static final float APP_LANGUAGE_QUALITY = .9f;
    private static final float SYSTEM_LANGUAGE_QUALITY = .8f;

    /**
     * @return The value that should go in the Accept-Language header.
     */
    @NonNull
    public static String getAcceptLanguage(@NonNull String wikiLanguageCode,
                                           @NonNull String appLanguageCode,
                                           @NonNull String systemLanguageCode) {
        String acceptLanguage = wikiLanguageCode;
        acceptLanguage = appendToAcceptLanguage(acceptLanguage, appLanguageCode, APP_LANGUAGE_QUALITY);
        acceptLanguage = appendToAcceptLanguage(acceptLanguage, systemLanguageCode, SYSTEM_LANGUAGE_QUALITY);
        return acceptLanguage;
    }

    @NonNull
    private static String appendToAcceptLanguage(@NonNull String acceptLanguage,
                                                 @NonNull String languageCode, float quality) {
        // If accept-language already contains the language, just return accept-language.
        if (acceptLanguage.contains(languageCode)) {
            return acceptLanguage;
        }

        // If accept-language is empty, don't append. Just return the language.
        if (acceptLanguage.isEmpty()) {
            return languageCode;
        }

        // Accept-language is nonempty, append the language.
        return String.format(Locale.ROOT, "%s,%s;q=%.1f", acceptLanguage, languageCode, quality);
    }

    private AcceptLanguageUtil() { }
}

File: app/src/main/java/org/wikipedia/language/AcceptLanguageUtil.kt
package org.wikipedia.language

import org.wikipedia.WikipediaApp
import java.util.Locale

object AcceptLanguageUtil {

    fun getAcceptLanguage(wikiLanguageCode: String, appLanguageCode: String, systemLanguageCode: String): String {
        val languageState = WikipediaApp.instance.languageState
        var quality = 1f

        // Compose a list of accepted languages, with descending quality values.
        // The first language is the most preferred, and has an implicit quality of 1.0.
        var acceptLanguage = wikiLanguageCode
        quality -= 0.1f

        if (wikiLanguageCode.startsWith("zh")) {
            // https://phabricator.wikimedia.org/T359822#10402218
            // For the specific case of Chinese, we provide the old-style language variant code,
            // in addition to the new-style BCP code. This seems to fix a tricky issue with the
            // mobile-html endpoint, and potentially others that use old-style variant codes.
            acceptLanguage = appendToAcceptLanguage(acceptLanguage, languageState.getBcp47LanguageCode(wikiLanguageCode), quality)
            quality -= 0.1f
        } else {
            acceptLanguage = languageState.getBcp47LanguageCode(wikiLanguageCode)
        }

        acceptLanguage = appendToAcceptLanguage(acceptLanguage, languageState.getBcp47LanguageCode(appLanguageCode), quality)
        quality -= 0.1f

        acceptLanguage = appendToAcceptLanguage(acceptLanguage, languageState.getBcp47LanguageCode(systemLanguageCode), quality)
        return acceptLanguage
    }

    private fun appendToAcceptLanguage(acceptLanguage: String, languageCode: String, quality: Float): String {
        if (acceptLanguage.contains(languageCode)) {
            return acceptLanguage
        }
        return if (acceptLanguage.isEmpty()) languageCode else String.format(Locale.ROOT, "%s,%s;q=%.1f", acceptLanguage, languageCode, quality)
    }
}
